### PR TITLE
Improve plugin catalog functionality by adding filtering

### PR DESCRIPTION
**Pull request by mattkelly**
#### What does this pull request accomplish?

Improve plugin catalog functionality by adding filtering

I'm mostly just interested in having these new APIs available for cluster create.
I'm not a huge fan of the APIs as they are right now, but I don't have any better ideas right now.

 #### What issue(s) does this fix?
* No issue, working towards #25 as always ;)

 #### Additional Considerations

 ### Testing

 #### Setup

 `make install`

 #### Instructions

```
csctl get pc --type csi
```
```
TYPE    IMPLEMENTATION          VERSIONS        KUBERNETES_COMPATIBILITY
CSI     azure                   0.1.0           1.10.x - 1.11.x
CSI     digitalocean            0.3.0           1.12.x - 1.12.x
CSI     digitalocean            0.2.0           1.10.5 - 1.11.x
CSI     digitalocean            0.1.1           1.10.x - 1.11.x
CSI     packet                  0.3.0           1.10.x - 1.11.x
CSI     google_cloud_platform   0.2.0-alpha     1.12.x - 1.12.x
CSI     google_cloud_platform   0.1.1           1.10.x - 1.11.x
CSI     google_cloud_platform   0.1.0           1.10.x - 1.11.x
```

```
csctl get pc --type csi --implementation digitalocean
```
```
TYPE    IMPLEMENTATION  VERSIONS        KUBERNETES_COMPATIBILITY
CSI     digitalocean    0.3.0           1.12.x - 1.12.x
CSI     digitalocean    0.2.0           1.10.5 - 1.11.x
CSI     digitalocean    0.1.1           1.10.x - 1.11.x
```

```
csctl get pc --type csi --implementation digitalocean --version 0.3.0
```
```
TYPE    IMPLEMENTATION  VERSIONS        KUBERNETES_COMPATIBILITY
CSI     digitalocean    0.3.0           1.12.x - 1.12.x
```

 ### Dependencies
* None

### CODE REVIEW

File: resource/plugin_catalog.go [hunk 1]
```diff
@@ -30,6 +30,65 @@ func PluginCatalog() *PluginsCatalog {
...
+	case "metrics":
+		pc.Metrics = []*types.PluginDefinition{def}
+	}
+
+	return NewPluginCatalog(&pc)
+}
+
+// NewPluginCatalogFromDefinitions constructs a new PluginsCatalog wrapping only
+// the given PluginDefinitions
+func NewPluginCatalogFromDefinitions(pluginType string, defs []*types.PluginDefinition) *PluginsCatalog {
```
Comment (mattkelly) [inline resource/plugin_catalog.go]: Refactor to combine with above.

**Comment by mattkelly** [comment]
Added some tests. This is now ready for review.

File: cmd/get_plugin_catalog.go [hunk 2]
```diff
@@ -1,24 +1,69 @@
...
 	Aliases: resource.PluginCatalog().Aliases(),
 
 	RunE: func(cmd *cobra.Command, args []string) error {
-		v, err := clientset.API().PluginCatalog().Get()
-		if err != nil {
-			return err
-		}
+		var plugins *resource.PluginsCatalog
+		switch {
+		case pluginType == "" && pluginImplementation == "" && pluginVersion == "":
```
Comment (ashleyschuett) [inline cmd/get_plugin_catalog.go]: should we be doing any validation on the types allowed? or are you fine with returning and error / empty plugins if they fill it in with non-valid information?

Comment (mattkelly) [reply cmd/get_plugin_catalog.go]: I figured we can rely on an API error for now. The error is quite nice:

```
csctl get pc --type asdf
```
```
Error: server responded with status 400: {"message":"Invalid params: \"type\" must be one of [autoscaler, cloud_controller_manager, cluster_management, cni, csi, logs, metrics]."}
```

I'm envisioning figuring out a better way to pull in the allowed and default parameters for various commands as part of https://github.com/containership/csctl/issues/66 (I know it sounds like just a documentation ticket, so I'll see if I can improve that).

There are many places where we're validating strings but then hardcoding the allowed strings in the usage, or not listing them at all. We should just figure out a better mechanism and use that moving forward.

**Pull request merged**